File: SystemLogQuery/LogDumper/LogDumper.m
#import "LogDumper.h"

NSString * const kLogDumperNotificationKey = @"com.briancroom.LogDumper.LogSomeStuff";
NSString * const kLogDumperNotificationAcknowledgementKey = @"com.briancroom.LogDumper.Acknowledgement";

static BOOL LogDumperDidReceiveAcknowledgement = NO;

static void LogDumperNotificationCallback(CFNotificationCenterRef center,
                                          void * observer,
                                          CFStringRef name,
                                          void const * object,
                                          CFDictionaryRef userInfo) {
    if ([kLogDumperNotificationKey isEqualToString:(__bridge NSString *)name]) {
        NSLog(@"** Dumping NSUserDefaults:\n\n%@\n\n", [NSUserDefaults standardUserDefaults].dictionaryRepresentation);
        CFNotificationCenterPostNotification(center, (__bridge CFStringRef)kLogDumperNotificationAcknowledgementKey, NULL, NULL, YES);
    } else {
        LogDumperDidReceiveAcknowledgement = YES;
    }
}

@implementation LogDumperTool

+ (void)load {
    NSLog(@"** Registering for log dumper notifications");

    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    for (NSString *name in @[kLogDumperNotificationKey, kLogDumperNotificationAcknowledgementKey]) {
        CFNotificationCenterAddObserver(center,
                                        (__bridge const void *)(self),
                                        LogDumperNotificationCallback,
                                        (__bridge CFStringRef)name,
                                        NULL,
                                        CFNotificationSuspensionBehaviorDeliverImmediately);
    }
}

+ (void)sendLogDumpRequest {
    LogDumperDidReceiveAcknowledgement = NO;

    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterPostNotification(center, (__bridge CFStringRef)kLogDumperNotificationKey, NULL, NULL, YES);
}

+ (void)sendLogDumpRequestAndWaitForAcknowledgement {
    [self sendLogDumpRequest];

    while (!LogDumperDidReceiveAcknowledgement) {
        [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    }
}

@end

File: Swift/SystemLogQuery/LogDumper/LogDumper.swift
import Foundation

@objc public class LogDumper: NSObject {
    private static let NotificationKey = "com.briancroom.LogDumper.LogSomeStuff"
    private static let NotificationAcknowledgementKey = "com.briancroom.LogDumper.Acknowledgement"

    private static var didReceiveAcknowledgement = false

    /// Calling this method posts a system-wide notification which will cause any other
    /// process which links with LogDumper to dump a bunch of internal state into the
    /// system log
    public static func sendLogDumpRequest() {
        didReceiveAcknowledgement = false

        let center = CFNotificationCenterGetDarwinNotifyCenter()
        CFNotificationCenterPostNotification(center, NotificationKey, nil, nil, true)
    }

    public static func sendLogDumpRequestAndWaitForAcknowledgement() {
        sendLogDumpRequest()

        while !didReceiveAcknowledgement {
            NSRunLoop.currentRunLoop().runUntilDate(NSDate(timeIntervalSinceNow: 0.1))
        }
    }

    @objc public static func registerForDumpNotification() {
        print("** Registering for log dumper notifications")

        let callBack: CFNotificationCallback = { center, _, name, _, _ in
            if name as String == LogDumper.NotificationKey {
                NSLog("** Dumping NSUserDefaults:\n\n%@\n\n", NSUserDefaults.standardUserDefaults().dictionaryRepresentation())
                CFNotificationCenterPostNotification(center, LogDumper.NotificationAcknowledgementKey, nil, nil, true)
            } else {
                LogDumper.didReceiveAcknowledgement = true
            }
        }

        let center = CFNotificationCenterGetDarwinNotifyCenter()
        for name in [NotificationKey, NotificationAcknowledgementKey] {
            CFNotificationCenterAddObserver(center, nil, callBack, name, nil, .DeliverImmediately)
        }
    }
}
